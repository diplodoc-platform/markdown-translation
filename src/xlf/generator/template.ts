import languages from '@cospired/i18n-iso-languages';
import countries from 'i18n-iso-countries';
import {XMLBuilder} from 'fast-xml-parser';

export type TemplateParameters = {
    source: LanguageLocale;
    target: LanguageLocale;
    skeletonPath: string;
    markdownPath: string;
};

export type LanguageLocale = {
    language: Language;
    locale: countries.Alpha2Code;
};

const languagesList = languages.langs();

export type Language = (typeof languagesList)[number];

const options = {
    format: true,
    ignoreAttributes: false,
    allowBooleanAttributes: true,
    suppressBooleanAttributes: false,
};

const builder = new XMLBuilder(options);

function generate(parameters: TemplateParameters) {
    if (!validParameters(parameters)) {
        throw new Error('invalid parameters');
    }

    const {source, target, markdownPath, skeletonPath} = parameters;

    const doc = {
        '?xml': {
            '@_version': '1.0',
            '@_encoding': 'UTF-8',
        },
        xliff: {
            '@_xmlns': 'urn:oasis:names:tc:xliff:document:1.2',
            '@_version': '1.2',
            file: {
                '@_original': markdownPath,
                '@_source-language': `${source.language}-${source.locale}`,
                '@_target-language': `${target.language}-${target.locale}`,
                '@_datatype': 'markdown',
                header: {
                    skeleton: {
                        'external-file': {
                            '@_href': skeletonPath,
                        },
                    },
                },
                body: {},
            },
        },
    };

    const rendered = builder.build(doc);

    let [before, after] = rendered.split('<body>');

    const indentation = before.length - before.trimEnd().length - 1;

    before += '<body>\n';
    after = ' '.repeat(indentation) + after;

    return {template: [before, after] as [string, string], indentation};
}

function validParameters(parameters: TemplateParameters) {
    const {source, target, skeletonPath, markdownPath} = parameters;

    const conditions = [
        ...[source, target].map(validLanguageLocale),
        ...[skeletonPath, markdownPath].map(Boolean),
    ];

    return conditions.reduce((a, v) => a && v, true);
}

function validLanguageLocale(parameters: LanguageLocale) {
    const {language, locale} = parameters;

    return languages.isValid(language) && countries.isValid(locale);
}

export {generate, validParameters};
export default {generate, validParameters};
