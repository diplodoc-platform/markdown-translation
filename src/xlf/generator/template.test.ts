import {generate, TemplateParameters} from './template';

describe('smoke', () => {
    it('works', () => {
        const parameters = {
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
            skeletonPath: 'file.skl.md',
            markdownPath: 'file.md',
        };

        generate(parameters);
    });
});

describe('validates parameters', () => {
    it('works with valid parameters', () => {
        const parameters = {
            markdownPath: 'file.md',
            skeletonPath: 'file.skl.md',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
        };

        generate(parameters);
    });

    it('throws on invalid parameters', () => {
        const invalidLocale = {
            markdownPath: 'file.md',
            skeletonPath: 'file.skl.md',
            source: {
                language: 'ru',
                locale: 'XX' as TemplateParameters['source']['locale'],
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
        };

        const invalidLanguage = {
            markdownPath: 'file.md',
            skeletonPath: 'file.skl.md',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'xx' as TemplateParameters['target']['language'],
                locale: 'US' as const,
            },
        };

        expect(() => generate(invalidLanguage)).toThrow();
        expect(() => generate(invalidLocale)).toThrow();
    });
});

describe('template', () => {
    it('generates xlf template', () => {
        const parameters = {
            markdownPath: 'file.md',
            skeletonPath: 'file.skl.md',
            source: {
                language: 'ru',
                locale: 'RU' as const,
            },
            target: {
                language: 'en',
                locale: 'US' as const,
            },
        };

        const {
            template: [before, after],
            indentation,
        } = generate(parameters);

        expect(indentation).toStrictEqual(4);
        expect(before).toMatchSnapshot();
        expect(after).toMatchSnapshot();
        expect(before + after).toMatchSnapshot();
    });
});
