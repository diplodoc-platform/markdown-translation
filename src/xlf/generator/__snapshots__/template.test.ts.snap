// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`template generates xlf template 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"file.md\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"file.skl.md\\"></external-file>
      </skeleton>
    </header>
    <body>
"
`;

exports[`template generates xlf template 2`] = `
"    </body>
  </file>
</xliff>
"
`;

exports[`template generates xlf template 3`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<xliff xmlns=\\"urn:oasis:names:tc:xliff:document:1.2\\" version=\\"1.2\\">
  <file original=\\"file.md\\" source-language=\\"ru-RU\\" target-language=\\"en-US\\" datatype=\\"markdown\\">
    <header>
      <skeleton>
        <external-file href=\\"file.skl.md\\"></external-file>
      </skeleton>
    </header>
    <body>
    </body>
  </file>
</xliff>
"
`;
